git-worktree/src/lib.rs

```diff
@@ -1 +1,95 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use anyhow::Result;
+use git_hash::oid;
+use git_index::{
+    entry::{Flags, Mode},
+    State,
+};
+use git_object::bstr::ByteSlice;
+use git_object::Data;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::time::SystemTime;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut State, path: P, mut find: Find, opts: Options) -> Result<()>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<Data<'a>>,
+{
+    let path = path.as_ref();
+    let paths = state
```

Comment:
I think it would be better to traverse `entries()` at first, and keep track of paths which later need their `stat` information be update by keeping only that information around. It would then be used to update the respective index entry with.